When statically parsing pyproject.toml, convert relative paths to absolute for use as file URIs

sirosen opened the pull request:
When statically parsing a `pyproject.toml`, we can encounter self-referential package usage. For example, an extra which names another extra on the same package.

The source for these requirements is the current package, but as a concrete requirement, not an abstract one. To make it concrete, we use a URI dependency link pointed at the package directory as a file-URI.

When the input path to `pyproject.toml` is relative, it needs to be made absolute for use as a URI. Failing to do so results in errors.

Because the `pip-compile` output of an such an extra depends concretely on the specific version of the package found in-tree, and not abstractly on the package name (without the URI), tests confirm that we see the rendered package directory as a file URI in `pip-compile` output.

<!--- Describe the changes here. --->

##### Contributor checklist

- [x] Included tests for the changes.
- [x] A change note is created in `changelog.d/` (see [`changelog.d/README.md`](https://github.com/jazzband/pip-tools/blob/main/changelog.d/#readme) for instructions) or the PR text says "no changelog needed".

##### Maintainer checklist

- [x] If no changelog is needed, apply the `skip-changelog` label.
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

webknjaz commented on changelog.d/2215.bugfix.md:
```diff
@@ -0,0 +1,2 @@
+Fix static parsing of `pyproject.toml` data when the `pyproject.toml` is
```
```suggestion
Fix static parsing of {file}`pyproject.toml` data when the {file}`pyproject.toml` is
```

webknjaz commented on tests/test_build.py:
```diff
@@ -95,6 +96,44 @@ def test_static_parse_valid(tmp_path):
     assert metadata.extras == ("baz",)
 
 
+@pytest.mark.parametrize("input_path_is_absolute", (True, False))
```
Could you add IDs here?

webknjaz commented on tests/test_cli_compile.py:
```diff
@@ -4044,3 +4044,63 @@ def test_second_order_requirements_relative_path_in_separate_dir(
     )
+
+
+@pytest.mark.parametrize("pyproject_path_is_absolute", (True, False))
```
IDs plz?

webknjaz commented on tests/test_cli_compile.py:
```diff
@@ -4044,3 +4044,63 @@ def test_second_order_requirements_relative_path_in_separate_dir(
+    pip_conf, runner, tmp_path, monkeypatch, pyproject_path_is_absolute
+):
+    """
+    Test that a `pyproject.toml` source file can use self-referential extras
```
Docstrings are usually parsed as RST:
```suggestion
    Test that a :file:`pyproject.toml` source file can use self-referential extras
```

webknjaz commented on tests/test_cli_compile.py:
```diff
@@ -4044,3 +4044,63 @@ def test_second_order_requirements_relative_path_in_separate_dir(
+    assert out.exit_code == 0
+    assert out.stdout == dedent(
+        f"""\
+        foo[ext1] @ {src_file.parent.absolute().as_uri()}
```
Is this really what we want? I don't think people would expect absolute paths from someone else's computer to end up in their lock files..

sirosen replied:
I agree, and it's why I was pretty hesitant at first. But also, that's the behavior from 7.4.1 -- it's just not tested.

In theory, someone could be using this in some controlled context like an image build and _wanting_ the concrete dependency path that this provides. So although I'm uncertain that it's right for the long term, I don't want to treat it as blocking the fix.

webknjaz replied:
Wait, did it always use absolute paths in 7.4.1? Not relative? I thought it'd conform to the shape of input that the 7.5.0 changelog promises.. Can we use `file://./rel/path/`?

webknjaz replied:
I suppose, we could postpone that for now. There might be a use case coverage missing for stripping the `@ http://..source` in some contexts.

webknjaz commented:
Oh, looks like Windows is unhappy about the delimiters again.

sirosen commented:
The thing I broke in the tests is treatment of `comes_from` data in this scenario. We're stringifying the path to the package dir here:
https://github.com/jazzband/pip-tools/blob/6e0ef336c9cad949753b6e1cc8bc3fbfecabd860/piptools/build.py#L91

It's easy to "fix" by just adding an `as_posix()` call there, but I can imagine this becoming inconsistent with the behavior when that information is rendered via some other path. I'm going to run some experiments later this week to try to figure out what the variations are on this.

maurerle commented:
Yes, this is exactly my issue. Would be very good to have this.
For now I either have to pin "pip-compile<7.5" or run pip-compile with "$PWD/pyproject.toml" to fix this.

sirosen commented:
I've rebased to keep this fresh with the latest fixes in `main` and added a new changelog note `changelog.d/2221.bugfix.md` to cover the new path normalization which happens.

With the `click`-related issue starting to impact users -- and now that I'm clear that it's not a `click` bug but an intentional change in behavior -- I'm wanting to push this forward and try to narrow what we put into 7.5.1 .

sirosen commented:
I think something is wrong with the codecov integration.
The report on uncovered lines doesn't make much sense:
https://app.codecov.io/gh/jazzband/pip-tools/pull/2221?dropdown=coverage

One thing I notice is that it seems to mark everything related to `piplowest` testing as uncovered, which suggests that somehow we're not producing distinct coverage data for those builds in a way that codecov understands.

I don't think it's at all related to this PR.

webknjaz reviewed (approved):
Looks good, but maybe add those change note refs before merging..

webknjaz commented on changelog.d/2215.bugfix.md:
@sirosen the issues didn't end up auto-closed. Next time, plz try to remember to mark the PR as closing them (via web UI or closes/resolves/fixes keywords).
